### PR TITLE
Rewrite the termination and solution statuses

### PR DESCRIPTION
**Pull request by joehuchette**
They are now "MOI light". This was semi-automated, so a translation mistake may have slipped through.

### CODE REVIEW

File: src/lp/dominated_column.jl [hunk 1]
```diff
@@ -43,8 +43,8 @@ function remove_dominated_column!(ps::PresolveData{T}, j::Int; tol::T=100 * sqrt
             ps.solution.s_upper .= zero(T)
 
             # Unbounded ray: xj = -1
-            ps.solution.primal_status = Sln_InfeasibilityCertificate
-            ps.solution.dual_status = Sln_Unknown
+            ps.solution.primal_status = INFEASIBILITY_CERTIFICATE
+            ps.solution.dual_status = NO_POINT
```
Comment (mtanneau) [inline src/lp/dominated_column.jl:47]: `NO_SOLUTION`

File: src/lp/dominated_column.jl [hunk 2]
```diff
@@ -100,8 +100,8 @@ function remove_dominated_column!(ps::PresolveData{T}, j::Int; tol::T=100 * sqrt
...
```
Comment (mtanneau) [inline src/lp/dominated_column.jl:104]: `NO_SOLUTION`

File: src/lp/empty_column.jl [hunk 3]
```diff
@@ -37,8 +37,8 @@ function remove_empty_column!(ps::PresolveData{T}, j::Int) where {T}
             ps.solution.s_upper .= zero(T)
 
             # Unbounded ray: xj = -1
-            ps.solution.primal_status = Sln_InfeasibilityCertificate
-            ps.solution.dual_status = Sln_Unknown
+            ps.solution.primal_status = INFEASIBILITY_CERTIFICATE
+            ps.solution.dual_status = NO_POINT
```
Comment (mtanneau) [inline src/lp/empty_column.jl:41]: `NO_SOLUTION`

File: src/lp/empty_column.jl [hunk 4]
```diff
@@ -69,8 +69,8 @@ function remove_empty_column!(ps::PresolveData{T}, j::Int) where {T}
             ps.solution.s_upper .= zero(T)
 
             # Unbounded ray: xj = 1
-            ps.solution.primal_status = Sln_InfeasibilityCertificate
-            ps.solution.dual_status = Sln_Unknown
+            ps.solution.primal_status = INFEASIBILITY_CERTIFICATE
+            ps.solution.dual_status = NO_POINT
```
Comment (mtanneau) [inline src/lp/empty_column.jl:73]: `NO_SOLUTION`

File: src/presolve.jl [hunk 5]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
...
+    ps.status == PRIMAL_INFEASIBLE && return ps.status
 
     # I. Remove all fixed variables, empty rows and columns
     # remove_fixed_variables!(ps)
     remove_empty_rows!(ps)
     remove_empty_columns!(ps)
 
     # TODO: check status for potential early return
-    ps.status == Trm_Unknown || return ps.status
+    ps.status == NOT_CALLED || return ps.status
```
Comment (mtanneau) [inline src/presolve.jl:375]: `IN_PROGRESS`

File: src/presolve.jl [hunk 6]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
```
Comment (mtanneau) [inline src/presolve.jl:364]: We need to switch `ps.status` to `IN_PROGRESS`

File: src/presolve.jl [hunk 7]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
...
+    ps.status == NOT_CALLED || return ps.status
 
     # Identify row singletons
     ps.row_singletons = [i for (i, nz) in enumerate(ps.nzrow) if ps.rowflag[i] && nz == 1]
 
     # II. Passes
     ps.updated = true
     npasses = 0  # TODO: maximum number of passes
-    while ps.updated && ps.status == Trm_Unknown
+    while ps.updated && ps.status == NOT_CALLED
```
Comment (mtanneau) [inline src/presolve.jl:383]: `IN_PROGRESS`

File: src/presolve.jl [hunk 8]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
...
         npasses += 1
         ps.updated = false
         @debug "Presolve pass $npasses" ps.nrow ps.ncol
 
         bounds_consistency_checks!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
Comment (mtanneau) [inline src/presolve.jl:389]: `IN_PROGRESS`

File: src/presolve.jl [hunk 9]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
...
         remove_empty_columns!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
Comment (mtanneau) [inline src/presolve.jl:391]: `IN_PROGRESS`

File: src/presolve.jl [hunk 10]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
...
 
 
         # Remove all fixed variables
         # TODO: remove empty variables as well
         remove_row_singletons!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
Comment (mtanneau) [inline src/presolve.jl:397]: `IN_PROGRESS`

File: src/presolve.jl [hunk 11]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
...
         remove_fixed_variables!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
Comment (mtanneau) [inline src/presolve.jl:399]: `IN_PROGRESS`

File: src/presolve.jl [hunk 12]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
...
 
         # Remove forcing & dominated constraints
         remove_row_singletons!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
Comment (mtanneau) [inline src/presolve.jl:403]: `IN_PROGRESS`

File: src/presolve.jl [hunk 13]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
...
         remove_forcing_rows!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
Comment (mtanneau) [inline src/presolve.jl:405]: `IN_PROGRESS`

File: src/presolve.jl [hunk 14]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
...
 
         # Remove free and implied free column singletons
         remove_row_singletons!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
Comment (mtanneau) [inline src/presolve.jl:409]: `IN_PROGRESS`

File: src/presolve.jl [hunk 15]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
...
         remove_free_column_singletons!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
Comment (mtanneau) [inline src/presolve.jl:411]: `IN_PROGRESS`

File: src/presolve.jl [hunk 16]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
...
 
         # TODO: remove column singleton with doubleton equation
 
         # Dual reductions
         remove_row_singletons!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
Comment (mtanneau) [inline src/presolve.jl:417]: `IN_PROGRESS`

File: src/presolve.jl [hunk 17]
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
...
         remove_dominated_columns!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
Comment (mtanneau) [inline src/presolve.jl:419]: `IN_PROGRESS`

File: src/presolve.jl [hunk 18]
```diff
@@ -507,8 +507,8 @@ function bounds_consistency_checks!(ps::PresolveData{T}) where {T}
             ps.solution.s_upper .= zero(T)
 
             # Farkas ray: y⁺_i = y⁻_i = 1 (any > 0 value works)
-            ps.solution.primal_status = Sln_Unknown
-            ps.solution.dual_status = Sln_InfeasibilityCertificate
+            ps.solution.primal_status = NO_POINT
```
Comment (mtanneau) [inline src/presolve.jl:510]: `NO_SOLUTION`

File: src/presolve.jl [hunk 19]
```diff
@@ -536,8 +536,8 @@ function bounds_consistency_checks!(ps::PresolveData{T}) where {T}
...
```
Comment (mtanneau) [inline src/presolve.jl:539]: `NO_SOLUTION`

File: src/presolve.jl [hunk 20]
```diff
@@ -584,7 +584,7 @@ If an empty column is created later, it is removed on the spot.
 function remove_empty_columns!(ps::PresolveData{T}) where {T}
     for j in 1:ps.pb0.nvar
         remove_empty_column!(ps, j)
-        ps.status == Trm_Unknown || break
+        ps.status == NOT_CALLED || break
```
Comment (mtanneau) [inline src/presolve.jl:587]: `IN_PROGRESS`

File: src/presolve.jl [hunk 21]
```diff
@@ -693,7 +693,7 @@ function remove_dominated_columns!(ps::PresolveData{T}) where {T}
 
     for (j, flag) in enumerate(ps.colflag)
         remove_dominated_column!(ps, j)
-        ps.status == Trm_Unknown || break
+        ps.status == NOT_CALLED || break
```
Comment (mtanneau) [inline src/presolve.jl:696]: `IN_PROGRESS`

File: test/lp/empty_column.jl [hunk 22]
```diff
@@ -38,25 +38,25 @@ function emtpy_column_tests(T::Type)
...
                 @test ps.ncol == 1
 
                 sol = ps.solution
-                @test sol.primal_status == MathOptPresolve.Sln_InfeasibilityCertificate
+                @test sol.primal_status == MOP.INFEASIBILITY_CERTIFICATE
                 @test sol.m == 0 && sol.n == 1
                 @test sol.x[1] > 0
             else
-                @test ps.status == MathOptPresolve.Trm_Unknown
+                @test ps.status == MOP.NOT_CALLED
```
Comment (mtanneau) [inline test/lp/empty_column.jl:59]: `REDUCED`

File: test/lp/empty_row.jl [hunk 23]
```diff
@@ -27,7 +27,7 @@ function empty_row_tests(T::Type)
     MathOptPresolve.remove_empty_row!(ps, 1)
 
     @test ps.updated
-    @test ps.status == MathOptPresolve.Trm_Unknown
+    @test ps.status == MOP.NOT_CALLED
```
Comment (mtanneau) [inline test/lp/empty_row.jl:30]: `REDUCED`

File: src/solution.jl [hunk 24]
```diff
@@ -18,7 +18,9 @@ mutable struct Solution{T}
     s_upper::Vector{T}
 
     Solution{T}(m, n) where {T} = new{T}(
-        m, n, Sln_Unknown, Sln_Unknown, false, false,
+        m, n,
+        NO_POINT, NO_POINT,
```
Comment (mtanneau) [inline src/solution.jl:22]: `NO_SOLUTION`

**Review by mtanneau — CHANGES_REQUESTED** [review]

**Comment by joehuchette** [comment]
OK, I have slightly restructed the ``TerminationStatus``es.

1. I renamed the enum to ``ModelStatus``, since it was also being used internally to track the status of the model as presolve progresses.
2. I lumped together ``NOT_CALLED``, ``IN_PROGRESS``, and ``REDUCED`` into ``NOT_INFERRED``. This means that presolve has not been able to infer the status of the model yet (e.g. optimality, primal infeasibility, etc.). IMO this is really what we're trying to track here, and it is true when presolve has not yet been called, when we are "in progress", and as a return type that means you still need to optimize over the inferred model.

I also added a little helper macro to get rid of all the status checks inside the presolve loop. I think it makes it a bit easier to read, but YMMV.

Let me know what you think.

**Review by mtanneau — APPROVED** [review]

**Comment by joehuchette** [comment]
Lol so it turns out ``presolve!`` isn't being tested, so my macro had a boneheaded escaping error. Fixed, and added a test for ``presolve!``.

**Pull request merged**